modules/rivrAnalyticsAdapter.js

```diff
@@ -0,0 +1,554 @@
+    } else {
+      let displayedImpression = iframe.contentDocument.getElementsByTagName('img').length > 0;
+      if (!displayedImpression) {
+        window.requestAnimationFrame(waitForElement);
+      } else {
+        specializedHandler(iframe);
+      }
+    }
+  }
+  waitForElement();
+};
+
+function addClickListener(iframe) {
+  iframe.contentDocument.addEventListener('click', reportClickEvent);
+}
+
+function impHandler(iframe) {
+  let timestamp = new Date().toISOString();
+  let requestId = generateUUID();
+  let adContainerId = iframe.parentElement.parentElement.id;
+  let impression = {
+    timestamp,
+    'request_id': requestId,
+    'tag_id': adContainerId
+  };
+  if (rivrAnalytics.context.queue) {
+    rivrAnalytics.context.queue.push(impression);
+  }
+}
+
+function addHandlers(bannersIds) {
+  bannersIds.forEach((bannerId) => {
+    addClickHandler(bannerId);
+    addDisplayedImpHandler(bannerId);
+  })
+};
+
+function fulfillAuctionObject() {
+  let newAuction = {
+    id: null,
+    timestamp: null,
+    at: null,
+    bcat: [],
+    imp: [],
+    app: {
+      id: null,
+      name: null,
+      domain: window.location.href,
+      bundle: null,
+      cat: [],
+      publisher: {
+        id: null,
+        name: null
+      }
+    },
+    site: {
+      id: null,
+      name: null,
+      domain: window.location.href,
+      cat: [],
+      publisher: {
+        id: null,
+        name: null
+      }
+    },
+    device: {
+      geo: {
+        city: null,
+        country: null,
+        region: null,
+        zip: null,
+        type: null,
+        metro: null
+      },
+      devicetype: getPlatformType(),
+      osv: null,
+      os: null,
+      model: null,
+      make: null,
+      carrier: null,
+      ip: null,
+      didsha1: null,
+      dpidmd5: null,
+      ext: {
+        uid: null
+      }
+    },
+    user: {
+      id: null,
+      yob: null,
+      gender: null,
+    },
+    bidResponses: [],
+    bidRequests: [],
+    'ext.rivr.optimiser': localStorage.getItem('rivr_should_optimise') || 'unoptimised',
+    modelVersion: localStorage.getItem('rivr_model_version') || null,
+    'ext.rivr.originalvalues': []
+  };
+  return newAuction;
+};
+
+function saveUnoptimisedParams() {
+  let units = rivrAnalytics.context.adUnits;
+  if (units) {
+    if (units.length > 0) {
+      let allUnits = connectAllUnits(units);
+      allUnits.forEach((adUnit) => {
+        adUnit.bids.forEach((bid) => {
+          let configForBidder = fetchConfigForBidder(bid.bidder);
+          if (configForBidder) {
+            let unOptimisedParamsField = createUnOptimisedParamsField(bid, configForBidder);
+            rivrAnalytics.context.auctionObject['ext.rivr.originalvalues'].push(unOptimisedParamsField);
+          }
+        })
+      });
+    }
+  }
+};
+
+function connectAllUnits(units) {
+  return units.reduce((acc, units) => {
+    units.forEach((unit) => acc.push(unit))
```

Comment:
`units` in this line becomes an object at this point since it's inside the `reduce`; using `forEach` returns an error.  Can you please review this area of the code?

Reply:
Sure, thanks!

Reply:
@jsnellbaker That function has been renamed to concatAllUnits() and I added a unit test to document what it does. The parameter `units` is an array of arrays `[bannersArray, videosArray]`, therefore we were using `reduce`.
The implementation has actually been changed to a more elegant one.

Here the test https://github.com/prebid/Prebid.js/pull/3202/files#diff-d54f743aba4b5542eeab77162622bea9R468